File: Handler/Handler Static Example 1/MainActivity.java
package com.example.myapplication;

import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import org.jetbrains.annotations.NotNull;

public class MainActivity extends AppCompatActivity {

    private TextView tvText;
    private static final String TAG = "TAG";
    private CustomHandler customHandler;
    private int range = 5;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //  Init Views
        initViews();
        //  Init Handler
        initHandler();
        //  Use
        onStartHandler();
    }

    private void initViews() {
        tvText = findViewById(R.id.tvText);
    }

    private void initHandler() {
        customHandler = new CustomHandler(this);
    }

    private void onStartHandler() {
        Thread t = new Thread(new Runnable() {
            public void run() {
                //  start iterate = 5
                for (int i = 0; i < range; i++) {
                    try {
                        Thread.sleep(1000);
                        customHandler.sendEmptyMessage(i +1);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        });
        t.start();
    }
    
      @Override
    protected void onDestroy() {
        super.onDestroy();
        if (customHandler != null) {
            customHandler = null;
        }
    }

    /*  Inner Static Class  */
    private static class CustomHandler extends android.os.Handler {

        private final java.lang.ref.WeakReference activity;

        CustomHandler(MainActivity activity) {
            this.activity = new java.lang.ref.WeakReference<>(activity);
        }

        @Override
        public void handleMessage(@NotNull android.os.Message msg) {
            Object o =  this.activity.get();
            MainActivity activity = (MainActivity) o;
            if (activity == null) {
                return;
            }
            Log.e(TAG, "Increment " + String.valueOf(msg.what));
            activity.tvText.setText("Increment " + String.valueOf(msg.what));
            //  Update UI

        }
    }
}

File: Kotlin/Functions/Handler/Handler Static Example 1/MainActivity.kt
package kz.snation.myapplication

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Message
import android.util.Log
import android.widget.TextView

class MainActivity : AppCompatActivity() {

    private lateinit var tvText: TextView
    companion object Variable {
        private val TAG = "TAG"
    }
    private var customHandler: CustomHandler? = null
    private var range: Int = 5

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //  Init Views
        initViews()
        //  Init Handler
        initHandler()
        //  Use
        onStartHandler()
    }

    private fun initViews() {
        tvText = findViewById(R.id.tvText)
    }

    private fun initHandler() {
        customHandler = CustomHandler(this)
    }

    private fun onStartHandler() {
        val t = Thread(Runnable {
             run {
                 //  start iterate = 5
                 for (i in 1..range) {
                     try {
                         Thread.sleep(1000)
                         customHandler!!.sendEmptyMessage(i)
                     } catch (e: Exception) {
                         e.printStackTrace()
                     }

                 }
             }
        })
        t.start()

    }

    override fun onDestroy() {
        super.onDestroy()
        if (customHandler != null) {
            customHandler!!.onDestroyHandler()
            customHandler = null
        }
    }

    /*  Inner Class  */
    @SuppressLint("HandlerLeak")
    private inner class CustomHandler(activity: MainActivity ): android.os.Handler() {

        private var activity: java.lang.ref.WeakReference<MainActivity>? = null

        init {
            this.activity = java.lang.ref.WeakReference<MainActivity>(activity)
        }

        override fun handleMessage(msg: Message) {
            val any: Any = this.activity!!.get()!!
            val activity: MainActivity = any as MainActivity
            if (activity == null) {
                return;
            }
            Log.e(TAG, "Increment " + msg.what)
            activity.tvText.text = "Increment " + msg.what
        }

        fun onDestroyHandler() {
            if (activity != null) {
                activity = null
            }
        }
    }
}
